yaksa: Add new package

raffenet opened the pull request:
Yaksa is primarily used as an MPICH dependency, but can be built as a standalone library.

raffenet committed:
yaksa: Add new package

haampie commented on var/spack/repos/builtin/packages/yaksa/package.py:
```diff
@@ -0,0 +1,37 @@
+    of contiguous segments) and flattening/unflattening datatypes into
+    process-portable formats.
+
+    Yaksa’s backend includes support for CPUs as well as different
```
```suggestion
    Yaksa's backend includes support for CPUs as well as different
```

raffenet committed:
yaksa: Add CUDA variant

raffenet commented:
Thanks for fixing up the initial commit @haampie. I squashed it an added a CUDA variant. This should be sufficient to use as an MPICH dependency.

alalazo commented on var/spack/repos/builtin/packages/yaksa/package.py:
```diff
@@ -0,0 +1,49 @@
+    depends_on('libtool',  type='build')
+    depends_on('m4',       type='build')
+
+    variant('cuda', default=False, description="CUDA backend support")
+    depends_on('cuda', when='+cuda')
```
Is it worth inheriting from `CudaPackage`?

raffenet replied:
Perhaps? Note that I also want to add Level Zero and HIP variants for Intel and AMD GPU support. Not sure if that influences the choice.

haampie replied:
The main advantage of inheriting from CudaPackage is getting conflicts for gcc and clang for concretization, that you otherwise get at compile time when including `cuda_runtime.h`: `error -- unsupported GNU version! gcc ... and up are not supported!`

raffenet replied:
Got it. I'll add it and the HIP and ZE variants to this PR.

raffenet replied:
Hm, is there any Level Zero package in Spack?

raffenet replied:
I added CudaPackage as suggested and also added a HIP variant. Since its not immediately clear how add a Level Zero dependency, I'm going to save that for future work. Removing the WIP tag, this should be OK to review again.

raffenet commented:
@alalazo @haampie are there any other issues with this package I should be looking into?

alalazo commented on var/spack/repos/builtin/packages/yaksa/package.py:
```diff
@@ -0,0 +1,56 @@
+    depends_on('m4',       type='build')
+
+    variant('cuda', default=False, description="CUDA backend support")
+    depends_on('cuda', when='+cuda')
```
These can be deleted, since now they are inherited from `CudaPackage`.

alalazo commented on var/spack/repos/builtin/packages/yaksa/package.py:
```diff
@@ -0,0 +1,56 @@
+    variant('cuda', default=False, description="CUDA backend support")
+    depends_on('cuda', when='+cuda')
+    variant('hip', default=False, description="HIP backend support")
+    depends_on('hip', when='+hip')
```
We have a similar `RocmPackage` that you can inherit from. The variant name is `rocm`, and it introduces a dependency on `hip` among other things.

alalazo commented on var/spack/repos/builtin/packages/yaksa/package.py:
```diff
@@ -0,0 +1,56 @@
+        spec = self.spec
+        config_args = []
+
+        if '+cuda' in spec:
+            config_args.append('--with-cuda={0}'.format(spec['cuda'].prefix))
+        else:
+            config_args.append('--without-cuda')
```
```suggestion
        config_args += self.with_or_without('cuda', activation_value='prefix')
```

raffenet commented:
> @alalazo I am trying to test this locally (`spack install yaksa +cuda`) but keep getting this error
> 
> ```
> ==> yaksa: Executing phase: 'configure'
> ==> Error: TypeError: 'NoneType' object is not iterable
> 
> /home/raffenet/spack/lib/spack/spack/build_systems/autotools.py:418, in configure:
>         415        and an appropriately set prefix.
>         416        """
>         417        options = getattr(self, 'configure_flag_args', [])
>   >>    418        options += ['--prefix={0}'.format(prefix)]
>         419        options += self.configure_args()
>         420
>         421        with working_dir(self.build_directory, create=True):
> ```
> 
> any clue what is happening?

Nevermind. I figured out I was missing a return statement in `configure_args`.

raffenet commented:
I squashed all my changes and dropped the HIP variant because there appear to be issues in our configuration scripts preventing it from building. Now this just adds Yaksa and the CUDA variant. Tagging @alalazo for another review.
